### ISSUE TITLE
missed fold, sext(fptosi x to i32) => fptosi x to i64

### ISSUE THREAD

**Issue by zhengyang92** [llvm/llvm-project#85268]
https://alive2.llvm.org/ce/z/bFW9vZ
online Alive timed out for this one, you'll need local alive2 with larger smt-to.

```llvm
define i64 @src(float %0) {
if.end27:
%1 = fptosi float %0 to i32
%2 = sext i32 %1 to i64
ret i64 %2

sink: ; No predecessors!
unreachable
}

define i64 @tgt(float %0) {
if.end27:
%1 = fptosi float %0 to i64
ret i64 %1

sink: ; No predecessors!
unreachable
}
```

no brainer, same pattern applies to fptoui.

**Comment by zhengyang92** [comment]
@jayfoad @regehr

**Comment by jayfoad** [comment]
My feeling is that we probably should not do this in IR because:
- It will not be a win on all targets. Float-to-i64 could be significantly more expensive than float-to-i32 on targets that do not have a single instruction for it (whereas sext i32 to i64 is a single fast instruction on every target we care about).
- Once you've done this transform, it's hard to undo it later.

Instead it could be done as a target-specific DAG combine or instruction selection pattern.

**Comment by nikic** [comment]
I agree that we should not do this in IR.

**Comment by zhengyang92** [comment]
@nikic @jayfoad agreed.

**Comment by miguelraz** [comment]
I'd like to work on this issue if possible 😀

**Comment by HendrikHuebner** [comment]
@miguelraz Are you still planning to work on this? Otherwise I'd like pick it up.

**Comment by miguelraz** [comment]
@HendrikHuebner Yes I am, sorry for licking the cookie for a bit - I'm just drowning in exams at the moment.

**Comment by medievalghoul** [comment]
Is this assuming that `sext` will always be the next instruction?

**Comment by arsenm** [comment]
> Is this assuming that `sext` will always be the next instruction?

The ordering / placement of instructions does not matter. The combine would be rooted at the sext and look at its operand

**Comment by c8ef** [comment]
Ping. @miguelraz
